src/coreclr/jit/emitloongarch64.cpp

```diff
@@ -0,0 +1,6797 @@
+            }
+            else
+            {
+                TempDsc* tmpDsc = codeGen->regSet.tmpFindNum(varNum);
+                vt              = tmpDsc->tdTempType();
+            }
+            if (vt == TYP_REF || vt == TYP_BYREF)
+                emitGCvarDeadUpd(adr + ofs, dst2 DEBUG_ARG(varNum));
+        }
+        // if (emitInsWritesToLclVarStackLocPair(id))
+        //{
+        //    unsigned ofs2 = ofs + TARGET_POINTER_SIZE;
+        //    if (id->idGCrefReg2() != GCT_NONE)
+        //    {
+        //        emitGCvarLiveUpd(adr + ofs2, varNum, id->idGCrefReg2(), *dp);
+        //    }
+        //    else
+        //    {
+        //        // If the type of the local is a gc ref type, update the liveness.
+        //        var_types vt;
+        //        if (varNum >= 0)
+        //        {
+        //            // "Regular" (non-spill-temp) local.
+        //            vt = var_types(emitComp->lvaTable[varNum].lvType);
+        //        }
+        //        else
+        //        {
+        //            TempDsc* tmpDsc = codeGen->regSet.tmpFindNum(varNum);
+        //            vt              = tmpDsc->tdTempType();
+        //        }
+        //        if (vt == TYP_REF || vt == TYP_BYREF)
+        //            emitGCvarDeadUpd(adr + ofs2, *dp);
+        //    }
+        //}
+    }
+
+#ifdef DEBUG
+    /* Make sure we set the instruction descriptor size correctly */
+
+    // size_t expected = emitSizeOfInsDsc(id);
+    // assert(sz == expected);
+
+    if (emitComp->opts.disAsm || emitComp->verbose)
+    {
+        code_t* cp = (code_t*)*dp;
+        while ((BYTE*)cp != dst)
+        {
+            emitDisInsName(*cp, (BYTE*)cp, id);
+            cp++;
+        }
+        // emitDispIns(id, false, dspOffs, true, emitCurCodeOffs(odst), *dp, (dst - *dp), ig);
+    }
+
+    if (emitComp->compDebugBreak)
+    {
+        // For example, set JitBreakEmitOutputInstr=a6 will break when this method is called for
+        // emitting instruction a6, (i.e. IN00a6 in jitdump).
+        if ((unsigned)JitConfig.JitBreakEmitOutputInstr() == id->idDebugOnlyInfo()->idNum)
+        {
+            assert(!"JitBreakEmitOutputInstr reached");
+        }
+    }
+#endif
+
+    /* All instructions are expected to generate code */
+
+    assert(*dp != dst);
+
+    *dp = dst;
+
+    return sz;
+}
+
+/*****************************************************************************/
+/*****************************************************************************/
+
+#ifdef DEBUG
+
+// clang-format off
+static const char* const RegNames[] =
+{
+    #define REGDEF(name, rnum, mask, sname) sname,
+    #include "register.h"
+};
+// clang-format on
+
+/****************************************************************************
+ *
+ *  Display the given instruction.
+ */
+
+// NOTE: At least 32bytes within dst.
+void emitter::emitDisInsName(code_t code, const BYTE* dst, instrDesc* id)
```

Comment:
I have a few concerns about this function:
1. One of the most important tools JIT developers when evaluating changes is "asm diffs", or viewing differences between a baseline compiler and a compiler with changes. To make these effective, uninteresting changes must be minimized. That especially means hiding actual addresses when `COMPlus_JitDiffableDasm=1` is set. All pointers need to be output using the `dspPtr` functions. Here, the address printing is spread out across hundreds of lines. Why not only print it in one location, at the top, and use dspPtr when printing it? Or, better, if DiffableDasm is set, don't print it at all, and reduce that verbosity. Nit: IMO, there's no reason to print `0x` prefix. The function emitDispInsAddr is for this.
2. Instead of having the instruction names as literal text, can't you look them up in the instruction table (using the `genInsName` table)?
3. It would be nice if the instruction name is aligned, and the instruction arguments are aligned. That is:
```
   0xff7d00c410   addi.d  sp, sp, -144
   0xff7d00c414   st.d    fp, sp, 0
   0xff7d00c418   st.d    ra, sp, 8
   0xff7d00c41c   ori     fp, sp, 0x0
   0xff7d00c420   addi.d  a3, fp, 32
   0xff7d00c424   addi.w  a4, zero, 5
   0xff7d00c428   st.d    zero, a3, 8
   0xff7d00c42c   st.d    zero, a3, 0
   0xff7d00c430   addi.d  a4, a4, -1
   0xff7d00c434   addi.d  a3, a3, 16
   0xff7d00c438   bne     a4, zero, 0xff7d00c428
   0xff7d00c43c   st.d    zero, a3, 0
   0xff7d00c440   st.d    a0, fp, 136
   0xff7d00c444   st.w    a1, fp, 132
   0xff7d00c448   st.d    a2, fp, 120
```
is better than:
```
   0xff7d00c410   addi.d  sp, sp, -144
   0xff7d00c414   st.d  fp, sp, 0
   0xff7d00c418   st.d  ra, sp, 8
   0xff7d00c41c   ori  fp, sp, 0x0
   0xff7d00c420   addi.d  a3, fp, 32
   0xff7d00c424   addi.w  a4, zero, 5
   0xff7d00c428   st.d  zero, a3, 8
   0xff7d00c42c   st.d  zero, a3, 0
   0xff7d00c430   addi.d  a4, a4, -1
   0xff7d00c434   addi.d  a3, a3, 16
   0xff7d00c438   bne  a4, zero, 0xff7d00c428
   0xff7d00c43c   st.d  zero, a3, 0
   0xff7d00c440   st.d  a0, fp, 136
   0xff7d00c444   st.w  a1, fp, 132
   0xff7d00c448   st.d  a2, fp, 120
```
4. You should print out opcode bytes at least optionally. That is what emitDispInsHex is for.

Reply:
@BruceForstall 
First of all , Thanks very much !

> 1. One of the most important tools JIT developers when evaluating changes is "asm diffs", or viewing differences between a baseline compiler and a compiler with changes. To make these effective, uninteresting changes must be minimized. That especially means hiding actual addresses when `COMPlus_JitDiffableDasm=1` is set. All pointers need to be output using the `dspPtr` functions.

(1) When I designed this, I'm not aware so many configuration. I focused on the practical needs for debugging.
And I know that this is not the best and can be improved further.
(2) e.g. I also met `asm diffs`, but we do it by vim tools which deleting some unintersting changes to look up the differences.
Of course, it's better that implementing this by JIT's switch values.

> 2. Instead of having the instruction names as literal text, can't you look them up in the instruction table (using the `genInsName` table)?

Of course there is not only one ways to implement this function.
Here is just for independent and easy for porting to release mode debugging.

> 3. It would be nice if the instruction name is aligned, and the instruction arguments are aligned. That is:

Yes, this can be implemented.
 
> 4. You should print out opcode bytes at least optionally. That is what emitDispInsHex is for.

Yes, at least optionally.

Reply:
@BruceForstall 
Could I push the `emitDisInsName` within the next patch ?

Reply:
yes

Reply:
I had pushed the patch for updating the format of the output instructions.